File: solutions/go/05/probability_dp/main.go
package main

import "fmt"

func probReach(n, maxFace int) float64 {
	dp := make([]float64, n+1)
	dp[0] = 1.0
	for i := 0; i < n; i++ {
		for f := 1; f <= maxFace; f++ {
			if i+f <= n {
				dp[i+f] += dp[i] / float64(maxFace)
			}
		}
	}
	return dp[n]
}

func main() {
	fmt.Printf("%.6f\n", probReach(6, 6))
}

File: solutions/python/05/probability_dp/main.py
def prob_reach(n: int, max_face: int = 6) -> float:
    dp = [0.0] * (n + 1)
    dp[0] = 1.0
    for i in range(n):
        for f in range(1, max_face + 1):
            if i + f <= n:
                dp[i + f] += dp[i] / max_face
    return dp[n]


def main() -> None:
    print(prob_reach(6, 6))


if __name__ == "__main__":
    main()
